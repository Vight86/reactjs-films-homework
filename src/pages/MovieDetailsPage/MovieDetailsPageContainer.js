import React, { useEffect } from 'react';
import propTypes from 'prop-types';
import {
  shallowEqual, useDispatch, useSelector,
} from 'react-redux';
import { useParams } from 'react-router-dom';

import { selectQuery } from '../../modules/query/index';
import { selectMovieDetails, loadMovieDetails } from '../../modules/movies/index';

import MovieDetailsPage from './MovieDetailsPage';
import Preloader from '../../components/Preloader/index';

const MovieDetailsPageContainer = ({ isModalOpened, handleModalToggle }) => {
  const dispatch = useDispatch();

  const { id } = useParams();
  const {
    basePath, apiKey, lang,
  } = useSelector(selectQuery);

  useEffect(() => {
    const fetchMovieDetailsUrl = `${basePath}/movie/${id}?api_key=${apiKey}&language=${lang}`;
    dispatch(loadMovieDetails(fetchMovieDetailsUrl));
    window.scrollTo(0, 0);
  }, [id]);

  const currentMovieDetails = useSelector(selectMovieDetails, shallowEqual);

  return (
    currentMovieDetails.title
      ? (
        <MovieDetailsPage
          isModalOpened={isModalOpened}
          handleModalToggle={handleModalToggle}
          id={id}
          currentMovieDetails={currentMovieDetails}
        />
      )
      : <Preloader />
  );
};

MovieDetailsPageContainer.propTypes = {
  isModalOpened: propTypes.bool.isRequired,
  handleModalToggle: propTypes.func.isRequired,
};

export default MovieDetailsPageContainer;
