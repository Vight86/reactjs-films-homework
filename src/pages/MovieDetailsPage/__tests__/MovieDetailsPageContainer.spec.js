import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MovieDetailsPageContainer from '../index';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn((f) => f()),
}));

let mockAppState = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(() => jest.fn()),
  useSelector: jest.fn().mockImplementation((callback) => callback(mockAppState)),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(() => ({ id: '2' })),
}));

window.scrollTo = jest.fn();

let renderer;
let result;

beforeEach(() => {
  mockAppState = {
    query: {
      basePath: 'http://testBasePath.com',
      apiKey: 'testApiKey',
      lang: 'testAlng',
      adult: false,
    },
    movies: {
      currentMovieDetails: {
        title: 'tested Title',
        backdroPath: '/tested',
        genres: [],
        runtime: 0,
        voteAverage: 0,
        overview: 'Tested overview',
      },
    },
  };

  renderer = new ShallowRenderer();
  renderer.render(
    <MovieDetailsPageContainer
      isModalOpened={false}
      handleModalToggle={jest.fn(() => 'test')}
    />,
  );
  result = renderer.getRenderOutput();
});

afterEach(() => {
  mockAppState = null;
  renderer = null;
  result = null;
});

describe('render MovieDetailsPageContainer component', () => {
  it('render component correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('renders Preloader while getting data', () => {
    mockAppState.movies.currentMovieDetails.title = '';

    renderer.render(
      <MovieDetailsPageContainer
        isModalOpened={false}
        handleModalToggle={jest.fn()}
      />,
    );
    expect(result).toMatchSnapshot();
  });
});
