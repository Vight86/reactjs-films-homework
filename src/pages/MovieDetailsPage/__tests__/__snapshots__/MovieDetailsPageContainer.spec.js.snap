// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MovieDetailsPageContainer component render component correctly 1`] = `
<MovieDetailsPage
  currentMovieDetails={
    Object {
      "backdroPath": "/tested",
      "genres": Array [],
      "overview": "Tested overview",
      "runtime": 0,
      "title": "tested Title",
      "voteAverage": 0,
    }
  }
  handleModalToggle={[MockFunction]}
  id="2"
  isModalOpened={false}
/>
`;

exports[`render MovieDetailsPageContainer component renders Preloader while getting data 1`] = `
<MovieDetailsPage
  currentMovieDetails={
    Object {
      "backdroPath": "/tested",
      "genres": Array [],
      "overview": "Tested overview",
      "runtime": 0,
      "title": "",
      "voteAverage": 0,
    }
  }
  handleModalToggle={[MockFunction]}
  id="2"
  isModalOpened={false}
/>
`;
